specs/sharding/beacon-chain.md

```diff
@@ -103,786 +87,558 @@ The following values are (non-configurable) constants used throughout the specif
-    # Once accepted by proposer, the signatures is the aggregate of both.
-    signature: BLSSignature
-```
+### Block processing
 
-### `PendingShardHeader`
+#### `is_intermediate_block_slot`
 
 ```python
-class PendingShardHeader(Container):
-    # The commitment that is attested
-    attested: AttestedDataCommitment
-    # Who voted for the header
-    votes: Bitlist[MAX_VALIDATORS_PER_COMMITTEE]
-    # Sum of effective balances of votes
-    weight: Gwei
-    # When the header was last updated, as reference for weight accuracy
-    update_slot: Slot
+def is_intermediate_block_slot(slot: Slot):
+    return slot % 2 == 1
 ```
 
-### `ShardBlobReference`
+### KZG
 
-Reference version of `ShardBlobHeader`, substituting the body for just a hash-tree-root.
+#### `hash_to_field`
 
 ```python
-class ShardBlobReference(Container):
-    slot: Slot
-    shard: Shard
-    # Builder of the data
-    builder_index: BuilderIndex
-    # Proposer of the shard-blob
-    proposer_index: ValidatorIndex
-    # Blob hash-tree-root for slashing reference
-    body_root: Root
+def hash_to_field(x: Container):
+    return int.from_bytes(hash_tree_root(x), "little") % MODULUS
 ```
 
-### `ShardProposerSlashing`
+#### `compute_powers`
 
 ```python
-class ShardProposerSlashing(Container):
-    slot: Slot
-    shard: Shard
-    proposer_index: ValidatorIndex
-    builder_index_1: BuilderIndex
-    builder_index_2: BuilderIndex
-    body_root_1: Root
-    body_root_2: Root
-    signature_1: BLSSignature
-    signature_2: BLSSignature
+def compute_powers(x: BLSFieldElement, n: uint64) -> List[BLSFieldElement]:
+    current_power = 1
+    powers = []
+    for i in range(n):
+        powers.append(BLSFieldElement(current_power))
+        current_power = current_power * int(x) % MODULUS
 ```
 
-### `ShardWork`
+#### `verify_kzg_proof`
 
 ```python
-class ShardWork(Container):
-    # Upon confirmation the data is reduced to just the commitment.
-    status: Union[                                                   # See Shard Work Status enum
-              None,                                                  # SHARD_WORK_UNCONFIRMED
-              AttestedDataCommitment,                                # SHARD_WORK_CONFIRMED
-              List[PendingShardHeader, MAX_SHARD_HEADERS_PER_SHARD]  # SHARD_WORK_PENDING
-            ]
-```
-
-## Helper functions
+def verify_kzg_proof(commitment: KZGCommitment, x: BLSFieldElement, y: BLSFieldElement, proof: KZGCommitment) -> None:
+    zero_poly = G2_SETUP[1].add(G2_SETUP[0].mult(x).neg())
 
-### Misc
+    assert (
+        bls.Pairing(proof, zero_poly)
+        == bls.Pairing(commitment.add(G1_SETUP[0].mult(y).neg), G2_SETUP[0])
+    )
+```
 
-#### `next_power_of_two`
+#### `interpolate_poly`
 
 ```python
-def next_power_of_two(x: int) -> int:
-    return 2 ** ((x - 1).bit_length())
+def interpolate_poly(xs: List[BLSFieldElement], ys: List[BLSFieldElement]):
+    """
+    Lagrange interpolation
+    """
+    # TODO!
```

Comment:
Maybe copy or take ideas from https://github.com/ethereum/research/blob/master/polynomial_reconstruction/poly_utils.py#L117 ?

Reply:
A polynomial class would probably make testing and compartimentalizing easier.

However from a spec point of view reduce+lambda and other Python idioms will need an explainer section (and we can explain range as well, in particular spelling that it's inclusive lower bound and exclusive upper bound).

Reply:
I will factor out polynomial operations as a class. However I am not sure about overloading BLS field element basic arithmetic.. For my own use this makes perfect sense to me, but there are probably some clients written in languages that do not support overloading and in that case, forgetting a modular reduction can have very subtle and potentially catastrophic consequences.

For polynomials this would be less of a concern, as bugs resulting from this can be more easily found and will typically cause clear compile time or run time errors.

Reply:
Thought about the polynomial class, but actually there aren't so many operations that this would make a significant difference for readability. In order to make things as clear as possible, I think it's best to just write out polynomial operations as well as field operations

Reply:
Thought about the polynomial class, but actually there aren't so many operations that this would make a significant difference for readability. In order to make things as clear as possible, I think it's best to just write out polynomial operations as well as field operations